### DISCUSSION TITLE
What do you want in the next major version?

### DISCUSSION THREAD

**Discussion by rez1dent3** [bavix/laravel-wallet#789] [Ideas]
There is already too much implemented in the wallet. Tell me what you would like to see inside the package?

**Comment by pourmami** [comment]
I would like to show withdraw and deposits history for each user separately. is it possible now? I don't see anything related in the docs

**Reply by rez1dent3** [reply]
It can be done now. I'll make an example for you later.

**Reply by rez1dent3** [reply]
If you want to withdraw transactions for all user wallets:
```php
$user->transactions()
```

If you want to withdraw transactions from the default wallet:
```php
$user->walletTransactions()
```

If you use multi-wallets and need to withdraw transactions for a specific wallet:
```php
$wallet->walletTransactions()
```

How to choose a specific wallet? https://bavix.github.io/laravel-wallet/#/new-wallet?id=how-to-get-the-right-wallet

**Reply by pourmami** [reply]
Thank you ♥

**Comment by DevMau00** [comment]
Native DB support for swap currency values, instead of having to load them thru service providers. 😁

**Reply by rez1dent3** [reply]
Please write your idea in more detail. Thank you.

If I understand you correctly, are you talking about replacing laravel-wallet-swap with a table inside a package that will already be filled somehow? Only the swap package is a common solution for everyone and its response is easily cached for the required time in any database.

**Reply by DevMau00** [reply]
The DB was just one way to work with the rates thru the controller, like if you want to fetch a rate from a previous rate you have on another table. Or another instance, if you get dynamic currency exchange rates from an external API from an user input on the front, and then needed to pass that value into the User thru a relationship so you could set the exchange first in the provider, and then on the controller you would need to make the deposit so the transaction could be created, yes?.

It would be awesome if you could set the rate at which the transaction its going to be exchanged from there. Skipping the step of setting the rate in the service provider first.

If there are any questions or if I didn't explain myself correctly let me know, so I can elaborate. Eng is not my main lang.

**Reply by DevMau00** [reply]
Something like:


// origin wallet / destination wallet / amount to exchange
$user->exchange(['USD' => 1.11],['EUR' => 0.91],100); 
//result 91 EUR 

Just an example. Dunno if it's dumb tho. Maybe theres already a way to do this and I'm just implementing it wrong. 😅

**Reply by rez1dent3** [reply]
Nowadays, you can’t manually specify a currency rate for a transaction (exchange), but you can fill it in globally or implement your custom service provider. 

An example implementation of a service provider can be found here: 
https://github.com/bavix/laravel-wallet-swap/blob/7109f1a2a5d947b1dc87ef9c49897cc2ff44d7ae/src/SwapExchangeService.php#L36

laravel-wallet-swap uses the florianv/laravel-swap library, and it can already be configured as you please. You can fill it through arrays, you can use API, or you can take data from tables.

Example via array:
https://github.com/bavix/laravel-wallet-swap/blob/7109f1a2a5d947b1dc87ef9c49897cc2ff44d7ae/tests/TestCase.php#L44

You can write a custom provider for the library and fill it from the database.

But perhaps I misunderstood you.

**Comment by rez1dent3** [comment]
#809

**Comment by rez1dent3** [comment]
add more events https://github.com/bavix/laravel-wallet/issues/819

**Comment by rez1dent3** [comment]
Documentation needs to be expanded #818

**Reply by rez1dent3** [reply]
https://github.com/bavix/laravel-wallet/pull/824

**Comment by gkmk** [comment]
@rez1dent3 is currency promotion in this update? I believe this to be something quite useful. 

The issue currently is that when listing **Transfers** (for reports or pagination), in order the get the actual amount and currency you need to load 2 more relationships (Transfer->Transaction->Wallet). My current solution is to extend the default models by adding these attributes inside transfer and transaction.

I know its kind of possible to update the **Meta** on transaction to keep this info but it makes indexing and querying a bit more difficult and it still requires the Transfer->Transaction relationship.

Am I using the package wrong or is this valid point?

**Reply by rez1dent3** [reply]
Hello. To be honest, I don't quite understand the context. Could you give an example of what you want to see and why.

**Reply by gkmk** [reply]
I want to build reports where i list all the transfers for one wallet.

What i have to do currently to access the Amount and Currency i need to do
```php
Number::currency($transfer->deposit->amount, $transfer->from->currency)
```

What i would like to do is avoid loading relationships and do this
```php
Number::currency($transfer->amount, $transfer->currency)
```

This is especially helpful when aggregating data as i think the `Transfer` model is the one that holds the most relevant data to display to the end customer.

Even maybe restructuring the `Transfer` model attributes 
Renaming `status` into `type` and `status_last` into `type_last`
And introducing new `status` attribute which would hold ENUM of something like `confirmed` in transactions.
The status ENUM would be something like `pending`, `confirmed`, `canceled`, `failed`

**Reply by rez1dent3** [reply]
Just an hour ago I added extra in a future version. Perhaps this will solve your problem.

https://github.com/bavix/laravel-wallet/pull/865

My proposal is this. All necessary data can be put into the new extra field and selected without additional queries. Yes, this raises the issue of data consistency, but this can also be solved, it seems to me.

**Comment by rez1dent3** [comment]
https://github.com/bavix/laravel-wallet/issues/955
